### DISCUSSION TITLE
Assembly entry

### DISCUSSION THREAD

**Discussion by joonicks** [PlummersSoftwareLLC/Primes#684] [General]
The poor state of this entry makes me sad.. sadly, Im horrible at using git so I dont really feel confident about contributing.

(x86_64) A simple loop unroll boosted performance by +50%

**Comment by rbergen** [comment]
As with any entry, I'm really interested to see the improvements that existing or new contributors can make. I therefore whole-heartedly invite you to put the effort into getting better at using git and then improving the Assembly entry to the absolute best of your ability.

**Comment by joonicks** [comment]
I was wondering, since malloc in libc is essentially a C function, in an assembly environment would the use or brk() and not unallocating after each run violate the faithfullness?

**Reply by rbergen** [reply]
Yes, it would. For a fair faithful comparison, certain base rules need to be universally applied even when it feels more "natural" not to, in a specific language. This has been extensively documented in earlier PRs and discussions.

That said, if you were to implement a runtime memory allocation/deallocation approach written in Assembly that replaces the one provided by libc, then we could certainly consider it.

**Comment by joonicks** [comment]
Ive come to realise that the optimizing task here isnt so much about making the code smaller but rather to figure out clever ways of minimizing the number of bit clearing operations. sieveLoop only runs 500 times, clearBitLoop is only entered 167 times.

**Reply by rbergen** [reply]
I'd phrase the challenge slightly differently. I'd say the challenge consists of two main parts: 
1. Indeed, not executing too many bit checking and clearing operations
2. Make the bit checking and clearing operations as cheap as possible, from the perspective of CPU cycles on the one hand, and memory access on the other. A relevant factor in the memory access part relates to clever use of the various levels of cache.

For both points, the rules for a base algorithm implementation do limit the optimizations that are allowed.

Looking at the current Assembly entries, there are a few factors at play when it comes to performance (I'm discussing my own code here, so I feel I'm allowed to make certain sweeping statements):
1. In more recent PRs, certain unrolling algorithms have been introduced in a number of languages that are fundamentally faster than the approach the Assembly entries use. These approaches achieve impressive speed improvements by minimizing the number of calculations and "slow" RAM references needed, at the expense of using more source code. 
2. The code has been written by a higher-level language programmer, with no actual background in (modern) CPU design. This means that there is practically a 100% chance that the current code frustrates instruction pipelining, breaks branch prediction, etc. Modern compilers for living languages have been written by people who do know the ins and outs of all that, which means they can optimize the machine code to insane levels. They do that at the cost of the disassembled machine code becoming impossible for me to read, but that's completely justified because legibility is not a goal. 
3. Sort of combining the previous two points, I have not spent any time on cache use optimizations. Other contributors, using other languages, have taken the time for that, and with clear performance results.

The long and short of it is that I wrote the Assembly entries because:
- my first submission to this project was an entry in MIX assembly (MIXAL), a computer for which I wrote an emulator that includes an assembler, which means I have a bit of a soft spot for Assembly variations
- no "modern CPU" Assembly entries were present yet, and I felt they absolutely had to be
- I'd never written anything in ARM Assembly, and I thought it would be fun to try

You'll note that "writing the absolutely fastest Assembly implementations possible" is not in that list. Which brings me back to my initial response to your post, which is that I welcome the contribution of others who do have the skills and do want to take the time to squeeze more out of the language. It sure is possible to.

**Reply by joonicks** [reply]
I just submitted my unrolled solution.

The way forward now to me seems to be figuring out how to calculate,
1. the length of the pattern produced,
2. how far to extend that pattern so its in place for the next pattern to be overlaid

for the factor 3, the pattern is 3 bytes, for factor 5 its 5 bytes
so initially the first pattern only needs to be extended 15? bytes to proceed with the algoritm, instead of 125000 bytes
writing 15 bytes instead of 125000 is one heck of a timesaver

its just very hard work to implement a new/experimental algorithm in assembly

**Reply by joonicks** [reply]
in fact, now im wondering if its possible to write out the pattern only once and then only extend it as needed for the new pattern

**Reply by joonicks** [reply]
in my head this works perfectly for factor 3 and 5.
factor 3 pattern (3 bytes) only needs to be written once
for factor 5, factor 3 needs to be continued six more times to a point they will be in sync, then factor 5 probably needs to be extended to that sync point.
after that my head starts hurting trying to imagine if the same algorithm pattern continues

**Reply by rbergen** [reply]
I have not yet looked at the code in your PR in detail, but looking at these comments it sounds to me like you may be calculating a bit mask in a (small) number of bytes for prime numbers 3 and 5, and then somehow projecting that mask onto (the rest of) the sieve. If that is indeed the case, the algorithm would no longer be considered the base algorithm. The definition of that algorithm includes that _"When clearing non-primes in the sieve (the second operation), **the algorithm clears all non-primes individually**, increasing the number with 2 * factor on each cycle."_

**Reply by joonicks** [reply]
If the last line of your comment was a rule, all unrolled loop solutions would be unfaitful/non base.

As for the "calculation" of a patters, thats a no. it started as an if-then-else exercise, but it turns out that for bit patters downsized to a byte, there are only two possibilities. for your original bitshift as it was writing dwords, that one has 8 different possibilities.
that the patterns repeats after 4 bytes is a happy coincidence/quirk of the math involved.
